File: SLR-backend/src/main/java/com/cnj/slrbackend/controller/PredictionController.java
package com.cnj.slrbackend.controller;

import com.cnj.slrbackend.utils.MultipartInputStreamFileResource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.io.IOException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.http.*;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.*;

@RestController
@RequestMapping("/python")
public class PredictionController {

    // 同级目录
    private final WebClient webClient;
    private final String WEIGHT_BASE = "SLR-S/weight/";
    private final String LOG_BASE = "SLR-S/logs/";

    @Autowired
    public PredictionController(WebClient webClient) {
        this.webClient = webClient;
    }

    /**
     * 1️⃣ 预测接口：调用 FastAPI
     */
    @PostMapping("/predict")
    public Mono<String> predict(
            @RequestParam(value = "file", required = false) MultipartFile file,
            @RequestParam("model") String model,
            @RequestParam("weight") String weight,
            @RequestParam("videoStyle") String videoStyle,
            @RequestParam("centercrop") boolean centercrop,
            @RequestParam(value = "video_path", required = false) String videoPath // 新增
    ) throws IOException {

        MultiValueMap<String, Object> formData = new LinkedMultiValueMap<>();
        formData.add("model", model);
        formData.add("weight", weight);
        formData.add("video_style", videoStyle);
        formData.add("centercrop", String.valueOf(centercrop));
        if (videoPath != null && !videoPath.isEmpty()) {
            formData.add("video_path", videoPath); // 测试集模式
        }
        if (file != null && !file.isEmpty()) {
            Resource fileResource = new MultipartInputStreamFileResource(file.getInputStream(), file.getOriginalFilename());
            formData.add("file", fileResource); // 上传/录制模式
        }

        return webClient.post()
                .uri("/predict/video")
                .contentType(MediaType.MULTIPART_FORM_DATA)
                .bodyValue(formData)
                .retrieve()
                .bodyToMono(String.class);
    }

    /**
     * 2️⃣ 权重文件列表
     */
    @GetMapping("/weights")
    public ResponseEntity<List<String>> getWeights(@RequestParam String model) {
        File weightDir = new File(WEIGHT_BASE + model);
        String[] files = weightDir.list((dir, name) -> name.endsWith(".pth") || name.endsWith(".pt"));
        if (files != null) {
            Arrays.sort(files, Comparator.reverseOrder());
            return ResponseEntity.ok(Arrays.asList(files));
        }
        return ResponseEntity.ok(Collections.emptyList());
    }

    /**
     * 3️⃣ 获取准确率曲线 CSV
     */
    @GetMapping("/metrics/acc")
    public ResponseEntity<List<String>> getAccCurve(@RequestParam String model) throws IOException {
        File csvFile = new File(LOG_BASE + "csv/" + model + "_acc.csv");
        List<String> lines = Files.readAllLines(csvFile.toPath(), StandardCharsets.UTF_8);
        return ResponseEntity.ok(lines);
    }

    /**
     * 4️⃣ 获取损失曲线 CSV
     */
    @GetMapping("/metrics/loss")
    public ResponseEntity<List<String>> getLossCurve(@RequestParam String model) throws IOException {
        File csvFile = new File(LOG_BASE + "csv/" + model + "_loss.csv");
        List<String> lines = Files.readAllLines(csvFile.toPath(), StandardCharsets.UTF_8);
        return ResponseEntity.ok(lines);
    }

    /**
     * 5️⃣ 网络结构 SVG
     */
    @GetMapping("/net/graph")
    public ResponseEntity<String> getNetworkGraph(@RequestParam String model) throws IOException {
        File svgFile = new File(LOG_BASE + "net/" + model + ".onnx.svg");
        String content = Files.readString(svgFile.toPath(), StandardCharsets.UTF_8);
        return ResponseEntity.ok(content);
    }

    /**
     * 6️⃣ 混淆矩阵 PNG（Base64编码）
     */
    @GetMapping("/metrics/confusion")
    public ResponseEntity<String> getConfusionMatrix(@RequestParam String model) throws IOException {
        File imageFile = new File(LOG_BASE + "confusion_matrix/" + model + "_confmat.png");
        byte[] bytes = Files.readAllBytes(imageFile.toPath());
        String base64 = Base64.getEncoder().encodeToString(bytes);
        return ResponseEntity.ok("data:image/png;base64," + base64);
    }
}

File: SLR-frontend/src/api/predictionController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/request'

/** 此处后端没有提供注释 GET /python/metrics/acc */
export async function getAccCurve(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getAccCurveParams,
  options?: { [key: string]: any }
) {
  return request<string[]>('/python/metrics/acc', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /python/metrics/confusion */
export async function getConfusionMatrix(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getConfusionMatrixParams,
  options?: { [key: string]: any }
) {
  return request<string>('/python/metrics/confusion', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /python/metrics/loss */
export async function getLossCurve(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getLossCurveParams,
  options?: { [key: string]: any }
) {
  return request<string[]>('/python/metrics/loss', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /python/net/graph */
export async function getNetworkGraph(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getNetworkGraphParams,
  options?: { [key: string]: any }
) {
  return request<string>('/python/net/graph', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /python/predict */
/** POST /python/predict 文件上传+参数 */
export async function predict(
  params: {
    model: string
    weight: string
    videoStyle: string
    centercrop: boolean
    videoPath?: string        // 新增，测试集模式下要传
  },
  file: File,
  options?: { [key: string]: any }
) {
  const formData = new FormData()
  formData.append('model', params.model)
  formData.append('weight', params.weight)
  formData.append('videoStyle', params.videoStyle)
  formData.append('centercrop', String(params.centercrop))
  if (params.videoPath) {
    formData.append('video_path', params.videoPath) // CSL测试集传入
  }
  if (file) {
    formData.append('file', file) // 上传/录制传入
  }
  return request<string>('/python/predict', {
    method: 'POST',
    data: formData,
    headers: {
      'Content-Type': 'multipart/form-data'
    },
    ...(options || {})
  })
}

/** 此处后端没有提供注释 GET /python/weights */
export async function getWeights(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getWeightsParams,
  options?: { [key: string]: any }
) {
  return request<string[]>('/python/weights', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}
